### PR TITLE
#7552 Allow adding a custom context for path management

### PR DESCRIPTION
**Pull request by viniciusxyz**
#7552 The idea of ​​this MR is to create the endpoints.all.context.path property that can be used to separate the management context-path from the server.context-path

With this change, the behavior of the paths is as follows:

**config:**

micronaut.server.context-path=/foo
endpoints.all.context-path=/fff
endpoints.all.path=/eee

**Now:**

Default app -> http://localhost:8080/foo
Management -> http://localhost:8080/fff/eee

**Before**

Default app -> http://localhost:8080/foo
Management -> http://localhost:8080/foo/eee

**config:**

micronaut.server.context-path=/foo
endpoints.all.path=/eee

**Now:**

Default app -> http://localhost:8080/foo
Management -> http://localhost:8080/foo/eee

**Before**

Default app -> http://localhost:8080/foo
Management -> http://localhost:8080/foo/eee

With the addition of this property there should be no change in behavior in relation to the current properties.

### CODE REVIEW

**Review by sdelamo — CHANGES_REQUESTED** [review]
This will be a breaking change.

**Comment by graemerocher** [comment]
not sure this makes sense to be honest. There should be a separate context path setting for endpoints in which case it wouldn't be breaking

**Comment by graemerocher** [comment]
To make this non-breaking IMO we should support `endpoints.all.context-path`

**Comment by viniciusxyz** [comment]
> To make this non-breaking IMO we should support `endpoints.all.context-path`

I will try to update MR over the weekend by adding this property.

**Comment by viniciusxyz** [comment]
> > To make this non-breaking IMO we should support `endpoints.all.context-path`
> 
> I will try to update MR over the weekend by adding this property.

I apologize, but due to Black Friday I had to work on the weekend and was unable to make the necessary changes, I will try to make it as soon as possible

**Comment by viniciusxyz** [comment]
@sdelamo I made the changes based on the comments made and instead of changing the behavior of **endpoints.all.path** the idea is to add the **endpoints.all.context-path** property, I am finishing the implementation, but I came across a difficulty that I still haven't been able to overcome, for some reason when adding the new property the **EndpointsFilter** filter is never executed, I'm investigating this and as soon as possible I'll add the documentation and remove the draft PR.

If you have any idea why a filter with the /** pattern is not filtering a request after adding a path, the tip would be very welcome.

File: management/src/main/java/io/micronaut/management/endpoint/EndpointDefaultConfiguration.java
Comment (sdelamo) [inline management/src/main/java/io/micronaut/management/endpoint/EndpointDefaultConfiguration.java:67]: Can you please this default to `null`

File: management/src/main/java/io/micronaut/management/endpoint/EndpointDefaultConfiguration.java
Comment (sdelamo) [inline management/src/main/java/io/micronaut/management/endpoint/EndpointDefaultConfiguration.java]: annotate the return type `@Nullable`

File: management/src/main/java/io/micronaut/management/endpoint/EndpointDefaultConfiguration.java
Comment (sdelamo) [inline management/src/main/java/io/micronaut/management/endpoint/EndpointDefaultConfiguration.java:131]: since contextPath must start with `/` and end with `/`, validate it here and throw a `ConfigurationException` if invalid.

File: management/src/main/java/io/micronaut/management/endpoint/processors/AbstractEndpointRouteBuilder.java
Comment (sdelamo) [inline management/src/main/java/io/micronaut/management/endpoint/processors/AbstractEndpointRouteBuilder.java]:
```suggestion
        var completePath = endpointDefaultConfiguration.getContextPath() == null
```

**Review by sdelamo — CHANGES_REQUESTED** [review]
Can you add a controller to the test to verify the controller is still accessible in the expected path?

Moreover, can you modify your test to test an endpoint which does not return 401. The Unauthorized makes the test confusing.

**Comment by sdelamo** [comment]
can you merge 4.3.x against your branch. I changed the base to be 4.3.x

File: management/src/main/java/io/micronaut/management/endpoint/EndpointDefaultConfiguration.java
Comment (viniciusxyz) [reply management/src/main/java/io/micronaut/management/endpoint/EndpointDefaultConfiguration.java:67]: Done

File: management/src/main/java/io/micronaut/management/endpoint/EndpointDefaultConfiguration.java
Comment (viniciusxyz) [reply management/src/main/java/io/micronaut/management/endpoint/EndpointDefaultConfiguration.java]: Done

File: management/src/main/java/io/micronaut/management/endpoint/processors/AbstractEndpointRouteBuilder.java
Comment (viniciusxyz) [reply management/src/main/java/io/micronaut/management/endpoint/processors/AbstractEndpointRouteBuilder.java]: Done

File: management/src/main/java/io/micronaut/management/endpoint/EndpointDefaultConfiguration.java
Comment (viniciusxyz) [reply management/src/main/java/io/micronaut/management/endpoint/EndpointDefaultConfiguration.java:131]: In fact, the comment is wrong, there is no need for it, I even believe that the comment about the context path for standard endpoints is also wrong.

**Comment by viniciusxyz** [comment]
> Can you add a controller to the test to verify the controller is still accessible in the expected path?
> 
> Moreover, can you modify your test to test an endpoint which does not return 401. The Unauthorized makes the test confusing.

As the test refers to the context path of the management endpoints and they are configured as sensitive by default, I have to guarantee that even with the change in context there is no change in the security model that the framework adds today, which is why I am following the applied testing model in the [EndpointsBasePathSpec](https://github.com/micronaut-projects/micronaut-core/blob/4.3.x/management/src/test/groovy/io/micronaut/management/endpoint/EndpointsBasePathSpec.groovy#L63) test which has a very similar objective, I still haven't been able to define why the security tests aren't passing, it's as if there was some trigger to activate the [EndpointsFilter](https://github.com/micronaut-projects/micronaut-core/blob/4.3.x/management/src/main/java/io/micronaut/management/endpoint/EndpointsFilter.java) that I can't see

**Comment by viniciusxyz** [comment]
> can you merge 4.3.x against your branch. I changed the base to be 4.3.x

I will do it tomorrow, I tried today, but there were a series of conflicts.

**Comment by viniciusxyz** [comment]
@sdelamo There seems to be a rule somewhere that when **micronaut.server.context-path** is defined the **EndpointsFilter** is only called when executing a path containing that context, in my unit test **" test routes with a server context path and endpoints context path"** when I define **micronaut.server.context-path** as **/** or remove the property, the test passes because I see it with a break point **EndpointsFilter** doFilter method is triggered in this case, for any other value I pass in **micronaut.server.context-path** the **EndpointsFilter**'s doFilter is never fired and therefore the endpoint is insecure . Can you tell me if there really is any rule like that? I'm confused because a filter that has the pattern `/**` should activate doFilter for any call. Or could I have misunderstood and really in Micronaut the Filter with pattern `/**` will always consider it as `{contextPath}/**` ?

**Comment by viniciusxyz** [comment]
@sdelamo The problem currently is that when defining **micronaut.server.context-path** and **endpoints.all.context-path** as in [this test](https://github.com/micronaut-projects/micronaut-core/pull/9820/files#diff-a7e03f43d7a65b07d3577352e534d84a6decf667a08c5939173bee737fabb0c4R64), **EndpointsFilter -> doFilter** is never called, this occurs because of this code snippet [ServerFilterRouteBuilder.java](https://github.com/micronaut-projects/micronaut-core/blob/4.2.x/router/src/main/java/io/micronaut/web/router/ServerFilterRouteBuilder.java#L68).

As a mental model, let’s think about the following:

Let's assume we have an application configured with Prometheus and the following properties:

- micronaut.server.context-path: /any
- endpoints.all.context-path: /test
- endpoints.prometheus.sensitive: true

1. A new enhancement of [AbstractEndpointRouteBuilder.java](https://github.com/micronaut-projects/micronaut-core/pull/9820/files#diff-295073bce4e6713906436635c42659f1c94c8fa58b80b719ea3534ede779e43f) creates the path as **/test/prometheus** ignoring the context path as it is what is desired.

2. A class [ServerFilterRouteBuilder](https://github.com/micronaut-projects/micronaut-core/blob/4.2.x/router/src/main/java/io/micronaut/web/router/ServerFilterRouteBuilder.java#L68) changes the default pattern `/**` of **EndpointsFilter** to ` /any/**`

This causes calls to the prometheus point, even if they are defined as sensitive true, to still be displayed without security. This is due to the filter's behavior. Is there any filter that I can use that does not have this behavior of append the contextPath?

**Comment by viniciusxyz** [comment]
@sdelamo Since I couldn't get the filter to ignore the ContextPath in the filters, I added a new property to the ServerFilter to establish whether the ContextPath should be added or not. With this and adding the option to not put ContextPath in the security filter now the solution works and all tests pass.

If you have any comments about the solution or need me to modify or test something, please let me know.

**Comment by altro3** [comment]
Any news about this improvement?

**Comment by viniciusxyz** [comment]
> Any news about this improvement?

I'm still waiting for approval :)

**Comment by altro3** [comment]
@graemerocher @sdelamo ping

**Comment by graemerocher** [comment]
it is scheduled for 4.5

**Comment by graemerocher** [comment]
Merged with https://github.com/micronaut-projects/micronaut-core/pull/10870 

Thanks for the contribution

**Pull request closed without merging**
